### ISSUE TITLE
bug(lineage): --symbol-lineage returns non-deterministic definition_count / reference_count across runs (25/25 → 18/32 → 28/22 for the same query)

### ISSUE THREAD

**Issue by aimasteracc** [aimasteracc/tree-sitter-analyzer#987]
From dogfood session on `develop` @ 35d0dac0 (2026-06-16). **Critical correctness bug.** Related to #978 / #981 (built-marker false-negative cascade).

**Symptom**

`--symbol-lineage <symbol>` returns **non-deterministic counts** across consecutive runs of the identical query. 5 back-to-back runs:

```
run 1 def: 25 ref: 25 risk_lvl: medium from_cache: False cache_inv: ast_index_stale
run 2 def: 18 ref: 32 risk_lvl: medium from_cache: False cache_inv: ast_index_stale
run 3 def: 28 ref: 22 risk_lvl: medium from_cache: False cache_inv: ast_index_stale
run 4 def: 28 ref: 22 risk_lvl: medium from_cache: False cache_inv: ast_index_stale
run 5 def: 28 ref: 22 risk_lvl: medium from_cache: False cache_inv: ast_index_stale
```

3 distinct count pairs (25/25, 18/32, 28/22) for the same query. Total def+ref is stable at 50, so it's the **classification that's flipping**, not the data set.

Repro (any project with built cache):

```
for i in 1 2 3 4 5; do
  uv run python -m tree_sitter_analyzer --symbol-lineage main --format json | \\
    python3 -c "import json,sys; d=json.loads(sys.stdin.read()); print(d['definition_count'], d['reference_count'])"
done
```

**Root cause (preliminary)**

`tree_sitter_analyzer/mcp/tools/symbol_lineage_tool.py:433-468` `_collect_definitions_and_refs` runs:

1. `execute_find_references` → `(definitions, references)`
2. `_reclassify_definition_like` → promotes definition-like references into `definitions`
3. `_enrich_references_with_callers` → adds call-graph callers to `references`

Steps 2 and 3 mutate the lists. The non-determinism is almost certainly in **step 2's content check** (`_line_looks_like_definition` at line 873 reads source files from disk) interleaved with **step 3's call-graph lookup**, which depends on the call-graph built marker. When the built marker is in a false-negative state (#978), the call-graph lookup falls through to in-memory construction whose iteration order is not stable.

`from_cache: False` and `cache_invalidated_reason: "ast_index_stale"` on every run confirms: the cache is being invalidated every call (because the staleness check distrusts the marker), so each call re-runs the full pipeline. Set/dict iteration order in `_reclassify_definition_like` or `_enrich_references_with_callers` then produces different promotion/enrichment outcomes.

**Why this matters**

1. **Agents cannot trust lineage data.** Two consecutive identical queries return different counts. An agent using lineage to decide "is this symbol safe to edit?" gets a different risk picture on retry.
2. **Risk level can flip.** In the repro above, `risk_lvl` happens to stay "medium" because total def+ref is stable, but `_assess_risk` keys off `def_count` and `ref_count` independently (lines 906-918). A symbol that lands on the boundary (`def_count=11` vs `def_count=9`) will flip between risk levels run-to-run.
3. **Caching is broken.** `from_cache: False` on every call means the lineage cache is useless when the index is "stale" — and per #978/#981, "stale" is itself a false-negative state. So the cache never warms, every call is slow (~16 s), and every call returns different data.
4. **Reproducibility for tests.** Any golden test that pins exact `definition_count` / `reference_count` will be flaky.

**Fix direction**

Two layers, both needed:

**A. (correctness) Make classification deterministic.**
- Audit `_reclassify_definition_like` (line 818) and `_enrich_references_with_callers` for set/dict iteration whose order influences outcomes. Sort inputs before iteration.
- If `_line_looks_like_definition` reads files (line 873), the file-read result should be deterministic — but if it's failing silently on some files (synthetic fixtures), the failure mode might be order-dependent. Make the failure mode explicit.
- Add a regression test: 10 consecutive `--symbol-lineage X` calls must return identical `definition_count` and `reference_count`.

**B. (root cause) Fix the staleness check so the cache warms.**
- This is the same root cause as #978 / #981: the built marker is false-negative, so `is_ast_index_stale` returns True even when the index is complete, so the lineage cache never hits.
- Fixing #978 will fix the perf symptom here (16 s → ~100 ms on cache hit), and fixing #981 will fix the misleading hint, but **classification determinism (layer A) must still be fixed independently** — the cache only hides the nondeterminism on warm hits; cold calls would still flip.

**Test gap**

Add to `tests/unit/test_agent_contracts.py`:

```python
def test_symbol_lineage_is_deterministic_across_runs(cli_runner, tmp_project_with_index):
    counts = set()
    for _ in range(10):
        result = cli_runner(["--symbol-lineage", "main", "--format", "json"])
        d = result.json
        counts.add((d["definition_count"], d["reference_count"]))
    assert len(counts) == 1, f"lineage non-deterministic: {counts}"
```

**Priority**

**P1** — silently-wrong data is worse than loudly-wrong data. This is the highest-severity finding from the 2026-06-16 dogfood session.

**Comment by aimasteracc** [comment]
**Verification after #1006 (P0 marker fix) merged:** #1006 correctly fixed the `call_graph_built()` false-negative (the real cache now reports built=True via the edges fallback, and #981's misleading hint is gone). **But this issue is NOT resolved by #1006** — re-tested on the warm 159K-edge cache: lineage counts still non-deterministic (run1 produced no JSON, run2 def=5/ref=45). This has an **independent root cause** beyond the marker. Keeping open.

**Comment by aimasteracc** [comment]
Resolved by #1009 (merged). Re-verified post-fix: 3 consecutive `--symbol-lineage execute` runs now return identical counts (def=6 ref=44) — deterministic. The non-determinism (25/25→18/32→28/22) was a side effect of the resolve-only refresh path's pathological per-file OR+LIKE DELETE SCAN (#990 root cause): each invocation observed a different partial edge/resolution state mid-refresh (or timed out). With the refresh now fast and convergent (#1009), the resolve pass completes deterministically before lineage reads it.
